stl/inc/span

```diff
@@ -0,0 +1,732 @@
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        if (_Off != 0) {
+            _STL_VERIFY(_Mydata, "cannot seek value-initialized span iterator");
+        }
+
+        if (_Off < 0) {
+#pragma warning(suppress : 4146) // unary minus operator applied to unsigned type, result still unsigned
+            _STL_VERIFY(_Myoff >= -static_cast<size_t>(_Off), "cannot seek span iterator before begin");
+        }
+
+        if (_Off > 0) {
+            _STL_VERIFY(_Mysize - _Myoff >= static_cast<size_t>(_Off), "cannot seek span iterator after end");
+        }
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        (void) _Off;
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+    }
+
+    constexpr _Span_iterator& operator+=(const difference_type _Off) noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _Verify_offset(_Off);
+        _Myoff += _Off;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        _Myptr += _Off;
+#endif // _ITERATOR_DEBUG_LEVEL
+
+        return *this;
+    }
+
+    _NODISCARD constexpr _Span_iterator operator+(const difference_type _Off) const noexcept {
+        _Span_iterator _Tmp{*this};
+        _Tmp += _Off;
+        return _Tmp;
+    }
+
+    constexpr _Span_iterator& operator-=(const difference_type _Off) noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        if (_Off != 0) {
+            _STL_VERIFY(_Mydata, "cannot seek value-initialized span iterator");
+        }
+
+        if (_Off > 0) {
+            _STL_VERIFY(_Myoff >= static_cast<size_t>(_Off), "cannot seek span iterator before begin");
+        }
+
+        if (_Off < 0) {
+#pragma warning(suppress : 4146) // unary minus operator applied to unsigned type, result still unsigned
+            _STL_VERIFY(_Mysize - _Myoff >= -static_cast<size_t>(_Off), "cannot seek span iterator after end");
+        }
+
+        _Myoff -= _Off;
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        _Myptr -= _Off;
+#endif // _ITERATOR_DEBUG_LEVEL
+
+        return *this;
+    }
+
+    _NODISCARD constexpr _Span_iterator operator-(const difference_type _Off) const noexcept {
+        _Span_iterator _Tmp{*this};
+        _Tmp -= _Off;
+        return _Tmp;
+    }
+
+    _NODISCARD constexpr difference_type operator-(const _Span_iterator& _Right) const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(
+            _Mydata == _Right._Mydata && _Mysize == _Right._Mysize, "cannot subtract incompatible span iterators");
+        return static_cast<difference_type>(_Myoff - _Right._Myoff);
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return _Myptr - _Right._Myptr;
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr reference operator[](const difference_type _Off) const noexcept {
+        return *(*this + _Off);
+    }
+
+    _NODISCARD constexpr bool operator==(const _Span_iterator& _Right) const noexcept {

```

Comment:
@CaseyCarter should advise us on spaceship-related matters.

Reply:
The rewrite rules have changed in the last few meetings as we've made `==` a magic defaultable operator in addition to and/or separately from `<=>`. `==` and `!=` no longer rewrite into `<=>` expressions, but `x != y` does rewrite into `!(x == y)`. If you default `<=>` you also get a defaulted `==`, but absent defaulting you need to implement both `==` and `<=>` to have all six of the equality and relational operators work for a type.

Consequently, we could implement just `==` and `<=>` here *if all of our supported compilers implement the necessary support*, but Clang doesn't support spaceship yet even on trunk. (It looks like EDG does, which I didn't realize before checking just now.) For things the standard doesn't depict directly (like most iterators) I think we should continue to implement all six operators as we would for C++17. For things the standard does depict directly as having only `==` and `<=>` - which is most of the standard library after merging WG21-P1614 - we should probably take the painful route of:
* implement `==`, 
* `#if _HAS_CXX20 && defined(__cpp_impl_three_way_comparison)`,  implement `<=>`
* otherwise, implement `!=`, `<`, `>`, `<=`, and `>=`.
With the expectation that we'll eventually be able to drop the conditional and the "other" operators for new C++20 things like `span`'s iterators when all our compilers support this part of C++20.

Reply:
@CaseyCarter you just made my head hurt. Feels like space sickness

Reply:
I think for now the answer is "no spaceship usage".

Reply:
"Let the person who implements WG21-P1614 worry about it" seems like a fine answer for now.